lib/cocoapods/sandbox/path_list.rb

```diff
@@ -47,15 +50,12 @@ def read_file_system
         unless root.exist?
           raise Informative, "Attempt to read non existent folder `#{root}`."
         end
-        root_length  = root.to_s.length + 1
         escaped_root = escape_path_for_glob(root)
-        paths = Dir.glob(escaped_root + '**/*', File::FNM_DOTMATCH).sort_by(&:upcase)
-        absolute_dirs  = paths.select { |path| File.directory?(path) }
-        relative_dirs  = absolute_dirs.map  { |p| p[root_length..-1] }
-        absolute_paths = paths.reject { |p| p == "#{root}/." || p == "#{root}/.." }
-        relative_paths = absolute_paths.map { |p| p[root_length..-1] }
-        @files = relative_paths - relative_dirs
-        @dirs  = relative_dirs.map { |d| d.gsub(/\/\.\.?$/, '') }.reject { |d| d == '.' || d == '..' } .uniq
+        absolute_paths = Pathname.glob(escaped_root + '**/*', File::FNM_DOTMATCH)
+        dirs_and_files = absolute_paths.reject { |path| path.basename.to_s =~ /^\.\.?$/ }
+        dirs, files = dirs_and_files.partition(&:directory?)
+        @dirs  = dirs.map{ |dir| dir.relative_path_from(root).to_s }.sort_by(&:upcase)
+        @files = files.map{ |file| file.relative_path_from(root).to_s }.sort_by(&:upcase)
```

Comment:
I gave it another go, see if you like this one :)

My main concerns with the existing implementation are:
- Relying on the root_length variable to do all operations.
- Not leveraging existing ruby classes for path manipulation.
- Too much checking for "." and ".." in various forms, each of those is removing some of the instances.

I won't push any more if you still prefer the old way, I just wanted to try and explain my thinking.